File: services/app/src/utils/providers/deezer.ts
import { sleep } from "../";
import { Album, MediaType, Playlist, Track } from "../medias";
import { SearchOptions, ProviderApi } from "../providers";
import { callProxy } from "../proxy";

// ------------------------------------------------------------------

const WWW_BASE = "https://www.deezer.com";
const RATE_LIMIT_DELAY = 5000; // ms

// ------------------------------------------------------------------

export type DeezerApiError = {
	code: number;
	message: string;
	type: string;
};

export type DeezerApiDataCollection<T> = {
	data: T[];
};

export type DeezerApiSearchResult<T> = {
	data: T[];
	total: number;
};

// ------------------------------------------------------------------

export type DeezerApiArtist = {
	id: number;
	name: string;
};

export type DeezerApiUser = {
	id: number;
	name: string;
};

// ------------------------------------------------------------------

export type DeezerApiAlbum = {
	error?: DeezerApiError;
	artist: DeezerApiArtist;
	cover_big: string;
	cover_small: string;
	id: number;
	title: string;
	tracks?: DeezerApiDataCollection<DeezerApiTrack>;
};

export type DeezerApiAlbumLight = {
	cover_big: string;
	cover_small: string;
	id: number;
	title: string;
};

export type DeezerApiPlaylist = {
	error?: DeezerApiError;
	creator?: DeezerApiUser;
	id: number;
	picture_big: string;
	picture_small: string;
	public: boolean;
	title: string;
	tracks?: DeezerApiDataCollection<DeezerApiTrack>;
	user?: DeezerApiUser;
};

export type DeezerApiTrack = {
	error?: DeezerApiError;
	album?: DeezerApiAlbumLight;
	artist: DeezerApiArtist;
	id: number;
	preview: string;
	readable: boolean;
	title: string;
};

// ------------------------------------------------------------------

const ConvertAlbum = (album: DeezerApiAlbum): Album => ({
	artist: {
		id: album.artist.id.toString(),
		name: album.artist.name,
		link: `${WWW_BASE}/artist/${album.artist.id}`
	},
	id: album.id.toString(),
	link: `${WWW_BASE}/album/${album.id}`,
	name: album.title,
	picture_big: album.cover_big,
	picture_small: album.cover_small,
	provider: "deezer",
	tracks:
		album.tracks !== void 0
			? album.tracks.data
					.filter(track => track.readable && track.preview)
					.map(track => ConvertTrack(track, album))
			: [],
	type: "album"
});

const ConvertPlaylist = (
	playlist: DeezerApiPlaylist,
	user: DeezerApiUser
): Playlist => ({
	id: playlist.id.toString(),
	link: `${WWW_BASE}/playlist/${playlist.id}`,
	name: playlist.title,
	picture_big: playlist.picture_big,
	picture_small: playlist.picture_small,
	provider: "deezer",
	tracks:
		playlist.tracks !== void 0
			? playlist.tracks.data
					.filter(track => track.readable && track.preview)
					.map(track => ConvertTrack(track, track.album!))
			: [],
	type: "playlist",
	user: {
		id: user.id.toString(),
		link: `${WWW_BASE}/profile/${user.id}`,
		name: user.name
	}
});

const ConvertTrack = (
	track: DeezerApiTrack,
	album: DeezerApiAlbumLight
): Track => ({
	album: {
		id: album.id.toString(),
		link: `${WWW_BASE}/album/${album.id}`,
		name: album.title,
		picture_big: album.cover_big,
		picture_small: album.cover_small
	},
	artist: {
		id: track.artist.id.toString(),
		link: `${WWW_BASE}/artist/${track.artist.id}`,
		name: track.artist.name
	},
	id: track.id.toString(),
	link: `${WWW_BASE}/track/${track.id}`,
	name: track.title,
	preview: track.preview,
	provider: "deezer",
	type: "track"
});

// ------------------------------------------------------------------

const DeezerApiImpl = (): ProviderApi => {
	const _call = <T>(
		path: string,
		params: { [key: string]: string }
	): Promise<T> => callProxy(`/deezer/${path}`, params);

	const _search = <T>(
		type: MediaType,
		query: string,
		options: SearchOptions
	) =>
		_call<DeezerApiSearchResult<T>>("search", {
			limit: options.limit.toString(),
			offset: options.offset.toString(),
			q: encodeURIComponent(query),
			type: type
		});

	const _load = async <T extends { error?: DeezerApiError }>(
		type: MediaType,
		ids: ReadonlyArray<string>
	): Promise<{
		delayedIds: string[];
		failedIds: string[];
		loadedMedias: T[];
	}> => {
		const delayedIds: string[] = [];
		const failedIds: string[] = [];
		const loadedMedias: T[] = [];
		await Promise.all(
			ids.map(async id => {
				const media = await _call<T>("load", { id, type });
				if (!media.error) {
					console.log("[Deezer] Loaded media", { type, id, media });
					loadedMedias.push(media);
					return;
				}
				if (media.error.code === 4) {
					delayedIds.push(id);
					return; // Quota limit exceeded, do not throw so that it will be retried later
				}
				console.error(
					`[Deezer] An error occured while loading ${type}:`,
					media.error
				);
				failedIds.push(id);
			})
		);
		return { delayedIds, failedIds, loadedMedias };
	};

	const _loadWithRetry = async <T extends { error?: DeezerApiError }>(
		type: MediaType,
		ids: ReadonlyArray<string>
	): Promise<T[]> => {
		let first = true;
		const res: T[] = [];
		let remainingIds = [...ids];
		while (remainingIds.length > 0) {
			if (!first) {
				console.debug(`[Deezer] Handling rate limit delayed batch...`, {
					remainingIds
				});
				await sleep(RATE_LIMIT_DELAY);
			}
			const { delayedIds, loadedMedias } = await _load<T>(
				type,
				remainingIds
			);
			res.push(...loadedMedias);
			remainingIds = delayedIds;
			first = false;
		}
		return res;
	};

	const searchAlbums = async (
		query: string,
		options: SearchOptions
	): Promise<Album[]> => {
		if (query.trim().length === 0) {
			return [];
		}
		return (
			await _search<DeezerApiAlbum>("album", query, options)
		).data.map(ConvertAlbum);
	};

	const searchPlaylists = async (
		query: string,
		options: SearchOptions
	): Promise<Playlist[]> => {
		if (query.trim().length === 0) {
			return [];
		}
		return (
			await _search<DeezerApiPlaylist>("playlist", query, options)
		).data
			.filter(playlist => playlist.public)
			.map(playlist => ConvertPlaylist(playlist, playlist.user!));
	};

	const searchTracks = async (
		query: string,
		options: SearchOptions
	): Promise<Track[]> => {
		if (query.trim().length === 0) {
			return [];
		}
		return (await _search<DeezerApiTrack>("track", query, options)).data
			.filter(track => track.readable && track.preview)
			.map(track => ConvertTrack(track, track.album!));
	};

	const loadAlbums = async (ids: ReadonlyArray<string>): Promise<Album[]> => {
		if (ids.length === 0) {
			return [];
		}
		console.debug("[Deezer] Loading albums...", { ids });
		const albums = await _loadWithRetry<DeezerApiAlbum>("album", ids);
		return albums.map(ConvertAlbum);
	};

	const loadPlaylists = async (
		ids: ReadonlyArray<string>
	): Promise<Playlist[]> => {
		if (ids.length === 0) {
			return [];
		}
		console.debug("[Deezer] Loading playlists...", { ids });
		const playlists = await _loadWithRetry<DeezerApiPlaylist>(
			"playlist",
			ids
		);
		return playlists.map(playlist =>
			ConvertPlaylist(playlist, playlist.creator!)
		);
	};

	const loadTracks = async (ids: ReadonlyArray<string>): Promise<Track[]> => {
		if (ids.length === 0) {
			return [];
		}
		console.debug("[Deezer] Loading tracks...", { ids });
		const tracks = await _loadWithRetry<DeezerApiTrack>("track", ids);
		return tracks.map(track => ConvertTrack(track, track.album!));
	};

	return {
		searchAlbums,
		searchPlaylists,
		searchTracks,
		loadAlbums,
		loadPlaylists,
		loadTracks
	};
};

// ------------------------------------------------------------------

let DEFAULT_IMPL: ProviderApi | null = null;

export const getDeezerApi = (): ProviderApi => {
	if (!DEFAULT_IMPL) {
		DEFAULT_IMPL = DeezerApiImpl();
	}
	return DEFAULT_IMPL;
};

File: services/api/deezer.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"strconv"
)

// ------------------------------------------------------------------

const API_BASE = "https://api.deezer.com";
const WWW_BASE = "https://www.deezer.com";

// ------------------------------------------------------------------

func callDeezer(url string) []byte {
	log.Printf("DEBUG - callDeezer: %s", url)

	res, err1 := http.Get(url)
	if err1 != nil {
		log.Fatal(err1)
	}
	defer res.Body.Close()

	body, err2 := ioutil.ReadAll(res.Body)
	if err2 != nil {
		log.Fatal(err2)
	}
	return body
}

type InitFunc func(length int) interface{}
type AppendFunc func(container interface{}, data []byte) interface{}

func loadDeezer(mediaType string, ids []string, init InitFunc, append AppendFunc) interface{} {
	channels := make([]interface{}, len(ids))
    for index, id := range ids {
		channels[index] = make(chan []byte)
		go func() {
			channels[index].(chan []byte) <- callDeezer(fmt.Sprintf("%s/%s/%s", API_BASE, mediaType, id))
		}()
	}
	medias := init(len(ids))
    for index, _ := range ids {
		medias = append(medias, <-channels[index].(chan []byte))
	}
	return medias
}


func searchDeezer(mediaType string, query string, limit int) []byte {
	return callDeezer(fmt.Sprintf("%s/search/%s?q=%s&limit=%d", API_BASE, mediaType, query, limit))
}

// ------------------------------------------------------------------

type DeezerApiError struct {
	Code int `json:"code"`
	Message string `json:"message"`
	Type string `json:"type"`
}

type DeezerApiAlbumLight struct {
	CoverBig string `json:"cover_big"`
	CoverSmall string `json:"cover_small"`
	ID int `json:"id"`
	Title string `json:"title"`
}

type DeezerApiArtist struct {
	ID int `json:"id"`
	Name string `json:"name"`
	PictureBig string `json:"picture_big"`
	PictureSmall string `json:"picture_small"`
}

type DeezerApiUser struct {
	ID int `json:"id"`
	Name string `json:"name"`
}

type DeezerApiAlbum struct {
	Error DeezerApiError `json:"error,omitempty"`
	Artist DeezerApiArtist `json:"artist"`
	CoverBig string `json:"cover_big"`
	CoverSmall string `json:"cover_small"`
	ID int `json:"ID"`
	Title string `json:"title"`
	Tracks DeezerApiTracksData `json:"tracks"`
}

type DeezerApiPlaylist struct {
	Error DeezerApiError `json:"error,omitempty"`
	Creator DeezerApiUser `json:"creator"`
	ID int `json:"id"`
	PictureBig string `json:"picture_big"`
	PictureSmall string `json:"picture_small"`
	Public bool `json:"public"`
	Title string `json:"title"`
	Tracks DeezerApiTracksData `json:"tracks"`
	User DeezerApiUser `json:"user"`
}

type DeezerApiTrack struct {
	Error DeezerApiError `json:"error,omitempty"`
	Album DeezerApiAlbumLight `json:"album"`
	Artist DeezerApiArtist `json:"artist"`
	Duration int `json:"duration"`
	ID int `json:"id"`
	Preview string `json:"preview"`
	Readable bool `json:"readable"`
	Title string `json:"title"`
}

type DeezerApiTracksData struct {
	Data []DeezerApiTrack `json:"data"`
}

// ------------------------------------------------------------------

type DeezerApiAlbumSearchResult struct {
	Data []DeezerApiAlbum `json:"data"`
	Total int `json:"total"`
}

type DeezerApiPlaylistSearchResult struct {
	Data []DeezerApiPlaylist `json:"data"`
	Total int `json:"total"`
}

type DeezerApiTrackSearchResult struct {
	Data []DeezerApiTrack `json:"data"`
	Total int `json:"total"`
}

// ------------------------------------------------------------------

func FilterAlbums(albums []DeezerApiAlbum) []Album {
	res := make([]Album, 0, len(albums))
    for _, album := range albums {
       	res = append(res, ConvertAlbum(album))
    }
    return res
}

func FilterPlaylists1(playlists []DeezerApiPlaylist) []Playlist {
	res := make([]Playlist, 0, len(playlists))
    for _, playlist := range playlists {
       	res = append(res, ConvertPlaylist(playlist, playlist.User))
    }
    return res
}

func FilterPlaylists2(playlists []DeezerApiPlaylist) []Playlist {
	res := make([]Playlist, 0, len(playlists))
    for _, playlist := range playlists {
       	res = append(res, ConvertPlaylist(playlist, playlist.Creator))
    }
    return res
}

func FilterTracks1(tracks []DeezerApiTrack) []Track {
	res := make([]Track, 0, len(tracks))
    for _, track := range tracks {
        if track.Readable && len(track.Preview) > 0 {
            res = append(res, ConvertTrack(track, track.Album))
        }
    }
    return res
}

func FilterTracks2(tracks []DeezerApiTrack, album DeezerApiAlbum) []Track {
	res := make([]Track, 0, len(tracks))
	albumLight := DeezerApiAlbumLight {
		CoverBig: album.CoverBig,
		CoverSmall: album.CoverSmall,
		ID: album.ID,
		Title: album.Title,
	}
    for _, track := range tracks {
        if track.Readable && len(track.Preview) > 0 {
            res = append(res, ConvertTrack(track, albumLight))
        }
    }
    return res
}

func ConvertAlbum(album DeezerApiAlbum) Album {
	return Album {
		Artist: Artist {
			ID: strconv.Itoa(album.Artist.ID),
			Name: album.Artist.Name,
			Link: fmt.Sprintf("%s/artist/%d", WWW_BASE, album.Artist.ID),
			PictureBig: album.Artist.PictureBig,
			PictureSmall: album.Artist.PictureSmall,
		},
		ID: strconv.Itoa(album.ID),
		Link: fmt.Sprintf("%s/album/%d", WWW_BASE, album.ID),
		PictureBig: album.CoverBig,
		PictureSmall: album.CoverSmall,
		Provider: "deezer",
		Title: album.Title,
		Tracks: FilterTracks2(album.Tracks.Data, album),
		Type: "album",
	}
}

func ConvertPlaylist(playlist DeezerApiPlaylist, user DeezerApiUser) Playlist {
	return Playlist {
		ID: strconv.Itoa(playlist.ID),
		Link: fmt.Sprintf("%s/playlist/%d", WWW_BASE, playlist.ID),
		PictureBig: playlist.PictureBig,
		PictureSmall: playlist.PictureSmall,
		Provider: "deezer",
		Title: playlist.Title,
		Tracks: FilterTracks1(playlist.Tracks.Data),
		Type: "playlist",
		User: User {
			ID: strconv.Itoa(user.ID),
			Link: fmt.Sprintf("%s/profile/%d", WWW_BASE, user.ID),
			Name: user.Name,
		},
	}
}

func ConvertTrack(track DeezerApiTrack, album DeezerApiAlbumLight) Track {
	return Track {
		Album: AlbumLight {
			ID: strconv.Itoa(album.ID),
			Link: fmt.Sprintf("%s/album/%d", WWW_BASE, album.ID),
			Title: album.Title,
			PictureBig: album.CoverBig,
			PictureSmall: album.CoverSmall,
		},
		Artist: Artist {
			ID: strconv.Itoa(track.Artist.ID),
			Name: track.Artist.Name,
			Link: fmt.Sprintf("%s/artist/%d", WWW_BASE, track.Artist.ID),
			PictureBig: track.Artist.PictureBig,
			PictureSmall: track.Artist.PictureSmall,
		},
		Duration: track.Duration,
		ID: strconv.Itoa(track.ID),
		Link: fmt.Sprintf("%s/track/%d", WWW_BASE, track.ID),
		Preview: track.Preview,
		Provider: "deezer",
		Title: track.Title,
		Type: "track",
	}
}

// ------------------------------------------------------------------

func loadAlbums(ids []string) <-chan []Album {
	x := func (length int) interface{} {
		return make([]DeezerApiAlbum, 0, length)
	}

	y := func (medias interface{}, data []byte) interface{} {
		media := DeezerApiAlbum {}
		err := json.Unmarshal(data, &media)
		if err != nil {
			log.Fatal(err)
		}
		if media.Error.Code == 4 {
			log.Print("Quota limit exceeded")
			return medias
		}
		if media.Error.Code != 0 {
			log.Print("An error occured while loading album: ", media.Error)
			return medias
		}
		return append(medias.([]DeezerApiAlbum), media)
	}

	channel := make(chan []Album)
	go func() {
		channel <- FilterAlbums(loadDeezer("album", ids, x, y).([]DeezerApiAlbum))
	}()
	return channel
}

func loadPlaylists(ids []string) <-chan []Playlist {
	x := func (length int) interface{} {
		return make([]DeezerApiPlaylist, 0, length)
	}

	y := func (medias interface{}, data []byte) interface{} {
		media := DeezerApiPlaylist {}
		err := json.Unmarshal(data, &media)
		if err != nil {
			log.Fatal(err)
		}
		if media.Error.Code == 4 {
			log.Print("Quota limit exceeded")
			return medias
		}
		if media.Error.Code != 0 {
			log.Print("An error occured while loading playlist: ", media.Error)
			return medias
		}
		return append(medias.([]DeezerApiPlaylist), media)
	}

	channel := make(chan []Playlist)
	go func() {
		channel <- FilterPlaylists2(loadDeezer("playlist", ids, x, y).([]DeezerApiPlaylist))
	}()
	return channel
}

func loadTracks(ids []string) <-chan []Track {
	x := func (length int) interface{} {
		return make([]DeezerApiTrack, 0, length)
	}

	y := func (medias interface{}, data []byte) interface{} {
		media := DeezerApiTrack {}
		err := json.Unmarshal(data, &media)
		if err != nil {
			log.Fatal(err)
		}
		if media.Error.Code == 4 {
			log.Print("Quota limit exceeded")
			return medias
		}
		if media.Error.Code != 0 {
			log.Print("An error occured while loading track: ", media.Error)
			return medias
		}
		return append(medias.([]DeezerApiTrack), media)
	}

	channel := make(chan []Track)
	go func() {
		channel <- FilterTracks1(loadDeezer("track", ids, x, y).([]DeezerApiTrack))
	}()
	return channel
};

// ------------------------------------------------------------------

func searchDeezerAlbums(query string, limit int) <-chan []Album {
	channel := make(chan []Album)
	go func() {
		defer close(channel)
		albums := DeezerApiAlbumSearchResult {}
		err := json.Unmarshal(searchDeezer("album", query, limit), &albums)
		if err != nil {
			log.Fatal(err)
		}
		channel <- FilterAlbums(albums.Data)
	}()
	return channel
}

func searchDeezerPlaylists(query string, limit int) <-chan []Playlist {
	channel := make(chan []Playlist)
	go func() {
		defer close(channel)
		playlists := DeezerApiPlaylistSearchResult {}
		err := json.Unmarshal(searchDeezer("playlist", query, limit), &playlists)
		if err != nil {
			log.Fatal(err)
		}
		channel <- FilterPlaylists1(playlists.Data)
	}()
	return channel
}

func searchDeezerTracks(query string, limit int) <-chan []Track {
	channel := make(chan []Track)
	go func() {
		defer close(channel)
		tracks := DeezerApiTrackSearchResult {}
		err := json.Unmarshal(searchDeezer("track", query, limit), &tracks)
		if err != nil {
			log.Fatal(err)
		}
		channel <- FilterTracks1(tracks.Data)
	}()
	return channel
}
